pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package rabbitmq
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const (
+	chansPrefix = "channels"
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
+	routingKey         = "mainfluxkey"
+	exchange           = "mainflux"
+	exchangeKind       = "fanout"
```

Comment:
Did you consider using `topic` kind? In that case, we have a queue per channel and the subtopic could be the routing key. This will have an impact on both publisher and subscriber implementation.

Reply:
I understand that we'll have to disassemble the topic we created before calling Publish/Subscribe methods, but it's worth trying it because topic manipulation is rather simple.

Reply:
From [this](https://www.cloudamqp.com/blog/part4-rabbitmq-13-common-errors.html) reference we need to have multiple queues. The current implementation I have pushed is that we have direct exchange rather than topic exchange as direct is faster than topic exchange. We will have a queue per subscription.
Another way to implement it is to have queues based on services such as a coap queue and an MQTT queue. But this makes us not have many queues.

Reply:
## Overview
RabbitMQ is somehow different to Nats and Kafka in that when a publisher is publishing a message on a topic they will publish the message to an exchange. The message will be published with a routing key. This routing key will direct the message to the different queues onto which the subscriber will consume the message. This is how the message flows
```
publisher -> exchange -> routingkey -> queue -> consumer -> subscriber


The publisher publishes the message to the broker.
The exchange receives the message being published and uses the routing key to route the message to a specific queue
The routing key is used to route the message to a specific queue
The queue temporarily stores messages for the consumer before consuming them
```

![image](https://user-images.githubusercontent.com/28790446/168829991-95a7977f-31ee-4289-952f-1ac121abc670.png)


## Exchange
When the exchange is declared we declare it with a specific kind. This means an exchange can either route messages based on topic, directly to queues or using message headers among other methods. We will be interested in the three routing techniques. The number of bindings increases the performance of topic routing. Fanout exchanges are very fast because there is no routing to process, but this changes if they are bound to a large number of queues. Direct exchanges are faster than topic exchanges if you don't need the wild card. We picked direct method as it strikes a balance between performance and reliability.

## Possible implementations

### One exchange one queue
- If you only have one queue, you'll end up repeating the logic of deciding which code should process the message, when RabbitMQ can do it for you with routing.
- RabbitMQ prefers many queues over a single large one
- This option is ruled out

### One exchange multiple queues
- This is the most viable solution currently
- This is because we will have single consumers on one queue
- The advantage of a single topic exchange is that you get the benefits of both a Direct Exchange and a Fanout Exchange depending on the binding key used.
- Further, configuration changes are easy to achieve and can often be done without disrupting any system processing at all

### Multiple exchanges one queue
- This option is ruled out based on having a single queue

### Multiple exchanges multiple queues
- This is the most preferred way of doing it as the system is highly decoupled. 
- If we cross the boundary and have many exchanges and many queues we are at the risk of the processes taking up more memory which may be significant but this really depends
- If the `msg.channel` is not provided the message won't be consumed as we can subscribe to an empty topic

## Wayforward
- Currently, we have implemented Direct as the exchange type. So if a consumer needs to subscribe to multiple topics we would change this implementation to the exchange having topic as the exchange type
- We have not explored exchange-to-exchange architectures

For more reference check out [this article](https://spring.io/blog/2011/04/01/routing-topologies-for-performance-and-scalability-with-rabbitmq/)